### ISSUE TITLE
Support stronger library code footprint reduction

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#11308]
A notorious problem with using OpenSSL on 'small' systems is its code size, in particular for the crypto lib.

Here is are some current figures for `libcrypto.a` on 64-bit Linux with the current master (of version 3.0.0-dev):
* 20 MB with `./config debug`
* 7.1 MB default, i.e, just `./config`
* 4.2 MB minimized (see below)
* 3.6 MB additionally prepending `CFLAGS=-Os`

Where minimizing has been done as follows:
```
./config no-shared no-dso no-pic no-aria no-async no-autoload-config \
  no-blake2 no-bf no-camellia no-cast no-chacha no-cmac \
  no-cmp no-cms no-comp no-ct no-des no-dgram no-dh no-dsa \
  no-dtls no-ec no-ec2m no-ecdsa no-engine no-filenames \ 
  no-gost no-idea no-ktls no-mdc2 no-md2 no-md4 no-multiblock \
  no-nextprotoneg no-ocsp no-ocb no-poly1305 no-psk \
  no-rc2 no-rc4 no-rc5 no-rmd160 no-scrypt no-seed no-siphash \
  no-siv no-sm2 no-sm3 no-sm4 no-srp no-srtp no-sse2 \
  no-ssl3 no-ssl3-method  no-threads no-ts no-ui-console no-whirlpool \
  no-asm -DOPENSSL_NO_SECURE_MEMORY -DOPENSSL_SMALL_FOOTPRINT
```

It looks like the code size cannot be reduced further without manually carving out code, right?
Do you see any chance of further reducing the library code footprint in a user-friendly way?

Counting the total size of `*.o` files per directory using
```
du -h -c -S --exclude='*.[cdhS]' --exclude='*.??*' crypto/ providers/ | sort -n
```
after minimizing as given above these are the sizes of the remaining code blobs:
```
8K	crypto/kdf
8K	crypto/perlasm
12K	crypto/txt_db
16K	crypto/async/arch
16K	crypto/buffer
16K	crypto/stack
20K	crypto/ec
20K	crypto/engine
20K	crypto/hmac
20K	crypto/lhash
20K	crypto/md5
20K	providers/implementations/asymciphers
20K	providers/implementations/keymgmt
28K	crypto/async
28K	providers/
28K	providers/common
32K	crypto/ess
32K	providers/implementations/macs
32K	providers/implementations/signature
40K	crypto/serializer
40K	providers/implementations/serializers
44K	crypto/dso
44K	crypto/ffc
48K	crypto/http
48K	crypto/sha
48K	providers/implementations/digests
56K	crypto/aes
60K	crypto/property
60K	crypto/ui
64K	crypto/err
72K	crypto/modes
76K	crypto/store
84K	crypto/conf
100K	crypto/pem
104K	providers/implementations/kdfs
116K	crypto/pkcs12
120K	crypto/pkcs7
152K	crypto/rand
192K	providers/implementations/ciphers
232K	crypto/objects
284K	crypto/bio
292K	crypto/rsa
308K	crypto/
312K	crypto/bn
628K	crypto/asn1
644K	crypto/evp
912K	crypto/x509
```

In particular, when not using certificates, ASN.1 (which of course would exclude a lot), RSA, random numbers, PKCS#7, PKCS#12, and/or PEM files, considerable size reductions should be achievable.

**Comment by mattcaswell** [comment]
I could envisage an "evp only" build that stripped out everything that wasn't pure crypto or necessary support code, so stuff like asn1, x509, pkcs7, pkcs12, bio etc. It would be quite an effort to get right though. I could also envisage a "tls only" build that only built those things absolutely necessary for libssl to work.

**Comment by levitte** [comment]
I envisage a "no legacy code" configuration, that should reduce matters quite a bit

**Comment by paulidale** [comment]
There was mention, a long while back, of building algorithms into individual providers.  By installing only those providers you want, the size could be more controlled.  When/if this will happen is unknown.

**Comment by levitte** [comment]
> When/if this will happen is unknown.

In that same discussion, I recall that we also said that this is in the far future, i.e. not for 3.0.

*But*, we might want to take a closer look at the possibility to create "bundle" providers, i.e. for someone to create a provider that bundles other providers together ...  there are subtleties in the current libcrypto \<-\> providder interface that make this impossible for the moment, but I think it's actually fairly easy to remedy.  I haven't explored that yet, though.

**Comment by paulidale** [comment]
Again after 3.0 -- we've got a lot on our collective plate already :)

**Comment by levitte** [comment]
Yeah well, if my intuition is correct, that kind of thing would break interface compatibility deeply, so pardon me, but I'm not letting that go quite yet.

**Comment by DDvO** [comment]
This issue is related to #7369 and #9982.

**Comment by erlmann** [comment]
> ... would break interface compatibility deeply, so pardon me, but I'm not letting that go quite yet.

@levitte could you explain how this would exactly break the interface compatibility? I would expect that, when I explicitly disable a functionality, I'm also not using the associated interfaces?

**Comment by levitte** [comment]
> @levitte could you explain how this would exactly break the interface compatibility? I would expect that, when I explicitly disable a functionality, I'm also not using the associated interfaces?

Sorry, that was my reply to @paulidale and the side discussion on making provider "bundles" possible.  So it has nothing to do with this PR specifically

**Comment by bernd-edlinger** [comment]
BTW: I had a PR for prime number generation speed up and code size reduction
here: #9387

**Comment by nhorman** [comment]
Marking as inactive, will be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
